Sources/App/Slack/Message.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  Message.swift
+//  App
+//
+//  Created by Allen Humphreys on 3/5/19.
+//
+
+import SlackKit
+
+// https://api.slack.com/messaging/managing
+// https://api.slack.com/messaging/sending#threading
+// https://api.slack.com/methods/chat.postMessage
+
+/// Represents a message from Slack. We could very easily create different types for
+/// incoming and outgoing messages to avoid having optional properties
+class Message {
+    let text: String
+    let channelID: ChannelID
+    let parent: String?
+
+    // I don't know that it makes sense to receive these?
+    var attachments = [Attachment]()
+
+    // Sender is only really useful for messages we've received
+    var sender: String? = nil
+    var timestamp: String? = nil
+
+    init(text: String, channelID: ChannelID, parent: String? = nil) {
+        self.text = text
+        self.channelID = channelID
+        self.parent = parent
+    }
+
+    public convenience init?(event: Event) {
+        guard let messageText = event.message?.text else { return nil }
+        guard let channelID = event.channel?.id else { return nil }
+
+        self.init(text: messageText, channelID: .init(id: channelID), parent: event.message?.threadTs)
+
+        sender = event.user?.id
+        timestamp = event.message?.ts
+    }
+
+    /// Constructs a new message in response to this one. If this message is in a thread,
+    /// the response will be in the same thread. If it is not, then the response won't be
+    func response(with text: String? = nil, attachments: [Attachment]? = nil) -> Message {
+        let response = Message(text: text ?? "", channelID: channelID, parent: parent)
+        attachments.map { response.attachments = $0 }
+        return response
+    }
+
+    /// Constructs a new message in response to this one. The response will be threaded on the original
+    func threadedResponse(with text: String? = nil, attachments: [Attachment]? = nil) -> Message {
+        let response = Message(text: text ?? "", channelID: channelID, parent: timestamp)
```

Comment:
Since the only difference looks to be this `parent: parent` vs `parent: timestamp`

Reply:
Yeah, so if you want to start a thread on a message, you use the message's `ts` as the parent. If the message you're responding to is already in a thread, you use it's `parent`

Reply:
See above. Do like the `if parent != nil` , a bit more clear as to what is going on then